Test monitor python example

BKnight760 opened the pull request:
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/<reponame>/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?
This pull request creates an example of the test monitor python client which is included in the Systemlink Python SDK.

### Why should this Pull Request be merged?
Users of the SystemLink Python SDK need examples of how to use the client APIs.

### What testing has been done?
The example has been tested locally on my SystemLink server and validated that the test result and steps are created as expected.

BKnight760 committed:
Add test monitor python client example for uploading test results and steps

BKnight760 committed:
Update readme

BKnight760 committed:
Update readme with SystemLink Python SDK dependency information

alexweav commented on python/testmonitor/results/README.md:
```diff
@@ -0,0 +1,20 @@
+This is an example of uploading test results to the SystemLink Test Monitor service.
+It simulates measuring the power output from a device and tests the measured power
+to ensure it is within a specified upper and lower limit.  The power is simulated using
+a the simple electrical equation `P=VI` (power=voltage*current).  In this example, a random
```
```suggestion
the simple electrical equation `P=VI` (power=voltage*current).  In this example, a random
```

alexweav commented on README.md:
```diff
@@ -10,5 +10,8 @@ The **jupyter** folder contains Jupyter notebook examples that exercise the diff
+The **dashboards** folder contains example layouts and tile configurations for customized use.
+- [testmonitor/OEE](./dashboards/testmonitor/OEE) - Overall Equipment Effectiveness
+
+The **python** folder contains examples of accessing the SystemLink server via python clients.
```
Really nitpicky:
```suggestion
The **python** folder contains examples of accessing the SystemLink server via Python clients.
```

alexweav commented on python/testmonitor/results/README.md:
```diff
@@ -0,0 +1,20 @@
+# SystemLink Test Monitor results example
+
+This is an example of uploading test results to the SystemLink Test Monitor service.
```
Is this description new, or is it taken from the C# one? If it's new, should we ask Jenna to review it?

BKnight760 replied:
This is new.  I'll add Jenna to the review.

alexweav commented on python/testmonitor/results/README.md:
```diff
@@ -0,0 +1,20 @@
+A top level result is created containing metadata about the overall test.
+
+The example sweeps across a range of input currents and voltages and takes measurements
+for each combination and stores a single measurement within each test step.  The test
```
```suggestion
for each combination. It then stores each single measurement within each test step.  The test
```

alexweav replied:
Any wording is fine here, as long as it breaks up the run-on.

alexweav commented on python/testmonitor/results/README.md:
```diff
@@ -0,0 +1,20 @@
+ultimately sets the status of the top-level test result.
+
+## Depenencies
+- The **NI SystemLink Python SDK** must be installed on the system using the NI Package Manager.  The package is available when installing the SystemLink Server or Client.
```
Should we note that it's installed to the python distribution under `C:\Program Files\National Instruments\...` just to avoid any confusion if they have Python already installed?

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -0,0 +1,224 @@
+from systemlink.clients.nitestmonitor.models.named_value_object import NamedValueObject
+
+
+def measure_power(current: float, voltage: float = 0) -> object:
```
`object` -> `Any`, which will also require `from typing import Any`

If we want to be more specific, we could also use `Tuple[float, List[Any], List[Any]]` or even specify `NamedValueObject`, but at that point it'll get really long so whatever you prefer.

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -0,0 +1,224 @@
+        power: float,
+        low_limit: float,
+        high_limit: float,
+        status: object) -> object:
```
We already imported `StatusObject`, so we could annotate this as `StatusObject`.

BKnight760 replied:
I wasn't sure how to do type annotations for custom classes.  Is it simply replacing `"object"` with `"StatusObject"`?

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -0,0 +1,224 @@
+    :param status:
+    :return: The step data used to create a test step.
+    """
+    step_status = status if status is not None else StatusObject(status_type='RUNNING')
```
```suggestion
    step_status = status if status else StatusObject(status_type='RUNNING')
```

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -0,0 +1,224 @@
+    return step_data
+
+
+def main():
```
This is optional since it's a `main`, but all other functions in this module have a docstring.

BKnight760 replied:
I thought about that, and had read this post: https://stackoverflow.com/questions/49782603/does-main-need-a-docstring which links to https://gist.github.com/dolph/39d8f70cab6afbac8c01.  Their justification was that main() should do what the module is designed to do.  So, in that case, the module documentation at the top is essentially the same thing that you'd put as documentation for main().  In the second link, they actually put main() as the first method in the file so it comes immediately after the module documentation.  Thoughts?

alexweav replied:
That seems reasonable to me. I suppose any docstring on it would always be super generic like "Entry point for the script" which is pretty unhelpful anyway because the use of `main` is obvious.

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -0,0 +1,224 @@
+
+            # Test the power measurement.
+            status = StatusObject(status_type='FAILED') if (power < low_limit or power > high_limit) else StatusObject(
+                status_type='PASSED')
```
Since the ternary is already so long that it wraps, do you think we could use a normal `if` here? e.g.

```
if power < low_limit or power > high_limit:
    status = StatusObject(status_type='FAILED')
else:
    status = StatusObject(status_type='PASSED')
```

alexweav replied:
This is optional for this PR, but we may want to consider running `black` over this, which will homogenize the formatting everywhere (`python -m black --line-length 120 .`). I suspect it would have changed this, but i'm not 100% sure.

(you also might need to install `black` manually, I don't think it's in the SL distribution...yet)

BKnight760 committed:
Update python/testmonitor/results/README.md

Co-authored-by: Alexander Weaver <alex.weaver@ni.com>

BKnight760 committed:
Update README.md

Co-authored-by: Alexander Weaver <alex.weaver@ni.com>

BKnight760 committed:
Update python/testmonitor/results/README.md

Co-authored-by: Alexander Weaver <alex.weaver@ni.com>

BKnight760 committed:
Update python/testmonitor/results/results.py

Co-authored-by: Alexander Weaver <alex.weaver@ni.com>

BKnight760 committed:
Lint results.py using black
Split ternary into simple if/else
Add install location of the python distro to the readme

BKnight760 committed:
Specify method return type for tuple.

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -45,24 +47,15 @@ def measure_power(current: float, voltage: float = 0) -> object:
-        low_limit: float,
-        high_limit: float,
-        status: object) -> object:
+def build_power_measurement_params(power: float, low_limit: float, high_limit: float, status: object) -> object:
```
A couple more uses of `object` type annotations here - could we either be more specific, or use `Any` from `typing`?

alexweav commented on python/testmonitor/results/results.py:
```diff
@@ -87,12 +80,13 @@ def build_power_measurement_params(
+    step_type: str,
+    inputs: object = None,
+    outputs: object = None,
+    parameters: object = None,
```
Same, `object` -> `Any`

BKnight760 committed:
Add type hints
